cmd/slinky/config/config.go

```diff
@@ -0,0 +1,269 @@
+		MaxPriceAge:    c.MaxPriceAge,
+		Providers:      providers,
+		Metrics:        c.Metrics,
+		Host:           c.Host,
+		Port:           c.Port,
+	}
+}
+
+func SetDefaults() {
+	setViperDefaultsForDataStructure("", DefaultOracleConfig())
+}
+
+func setViperDefaultsForDataStructure(keyPrefix string, config interface{}) {
+	val := reflect.ValueOf(config)
+	typ := val.Type()
+
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Field(i)
+		jsonFieldTag := typ.Field(i).Tag.Get("json")
+
+		// the fully-qualified key for this field
+		fullKey := keyPrefix + jsonFieldTag
+
+		switch field.Kind() {
+		case reflect.Struct:
+			// set viper defaults for struct via recursion
+			setViperDefaultsForDataStructure(fullKey+jsonFieldDelimiter, field.Interface())
+		case reflect.Map:
+			// set viper defaults for map
+			for _, key := range field.MapKeys() {
+				setViperDefaultsForDataStructure(
+					fullKey+jsonFieldDelimiter+key.String()+jsonFieldDelimiter,
+					field.MapIndex(key).Interface(),
+				)
+			}
+		default:
+			viper.SetDefault(fullKey, field.Interface())
+		}
+	}
+
+	// set the environment prefix
+	viper.SetEnvPrefix(SlinkyConfigEnvironmentPrefix)
+	viper.AutomaticEnv()
+	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+}
+
+func GetLegacyOracleConfig(path string) (config.OracleConfig, error) {
+	viper.SetConfigFile(path)
+	viper.SetConfigType("json")
+
+	if err := viper.ReadInConfig(); err != nil {
+		return config.OracleConfig{}, err
+	}
+
+	var cfg config.OracleConfig
+	if err := viper.Unmarshal(&cfg); err != nil {
+		return config.OracleConfig{}, err
+	}
+
+	return cfg, cfg.ValidateBasic()
+}
+
+// ReadOracleConfigFromFile reads a config from a file and returns the config.
+func ReadOracleConfigWithOverrides(path string, marketMapProvider string) (config.OracleConfig, error) {
+	// if the path is non-nil read data from a file\
+	SetDefaults()
+	if path != "" {
+		viper.SetConfigFile(path)
+		viper.SetConfigType("json")
+
+		if err := viper.ReadInConfig(); err != nil {
+			return config.OracleConfig{}, err
+		}
+	}
+
+	cfg, err := oracleConfigFromViper()
+	if err != nil {
+		return config.OracleConfig{}, err
+	}
+
+	// filter the market-map providers
+	if _, ok := constants.MarketMapProviderNames[marketMapProvider]; !ok {
+		return config.OracleConfig{}, fmt.Errorf("market map provider %s not found", marketMapProvider)
+	}
+
+	// filter the unused market-map providers
+	for name, provider := range cfg.Providers {
+		if provider.Type == mmtypes.ConfigType {
+			if name != marketMapProvider {
+				delete(cfg.Providers, name)
+			}
+		}
+	}
+
+	return cfg.ToLegacy(), nil
+}
+
+// oracleConfigFromViper unmarshals an oracle config from viper, validates it, and returns it.
+func oracleConfigFromViper() (OracleConfig, error) {
+	var cfg OracleConfig
+	if err := viper.Unmarshal(&cfg); err != nil {
+		return OracleConfig{}, err
+	}
+
+	// for each api-provider, we'll have to manually fill the endpoints
```

Comment:
This is only allowing us to use env variables to modify the api endpoints? Do we want to be able to modify things like host, ws, and other?

Reply:
discussed offline